src/docker.c

```diff
@@ -0,0 +1,1137 @@
+    sfree(stats->memory_stats);
+    return;
+  }
+
+  unsigned long *result_ptrs[] = {
+    &(stats->memory_stats->usage),
+    &(stats->memory_stats->limit),
+  };
+
+  assert(STATIC_ARRAY_SIZE(MEMORY_RESPONSE_KEYS) ==
+         STATIC_ARRAY_SIZE(result_ptrs));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(MEMORY_RESPONSE_KEYS); i++) {
+    extract_value(memory_node, MEMORY_RESPONSE_KEYS[i], result_ptrs[i]);
+  }
+  compute_memory_stats(stats->memory_stats);
+}
+
+// Extracts stats for a given interface.
+static void get_interface_stats(yajl_val interface_node,
+    interface_stats_t * stats) {
+  //Network Stats
+  unsigned long *rx_result_ptrs[] = {
+    &(stats->rx_bytes),
+    &(stats->rx_packets),
+    &(stats->rx_errors),
+  };
+
+  unsigned long *tx_result_ptrs[] = {
+    &(stats->tx_bytes),
+    &(stats->tx_packets),
+    &(stats->tx_errors),
+  };
+
+  assert(STATIC_ARRAY_SIZE(INTERFACE_KEYS) == STATIC_ARRAY_SIZE(rx_result_ptrs));
+  assert(STATIC_ARRAY_SIZE(INTERFACE_KEYS) == STATIC_ARRAY_SIZE(tx_result_ptrs));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(INTERFACE_KEYS); i++) {
+    int len = strlen(INTERFACE_KEYS[i]) + 4; //tx_{bytes}\0
+    char *tx_key = (char *) calloc(len, sizeof(char));
+    char *rx_key = (char *) calloc(len, sizeof(char));
+    if (tx_key == NULL || rx_key == NULL) {
+      ERROR("docker: get_metrics_for_container: malloc failed.\n");
+    }
+    ssnprintf(tx_key, len, "tx_%s", INTERFACE_KEYS[i]);
+    ssnprintf(rx_key, len, "rx_%s", INTERFACE_KEYS[i]);
+    extract_value(interface_node, tx_key, tx_result_ptrs[i]);
+    extract_value(interface_node, rx_key, rx_result_ptrs[i]);
+    sfree(tx_key);
+    sfree(rx_key);
+  }
+}
+
+// Retrieves Network statistics for each networking interface from the STATS API
+// response.
+static void get_network_stats(yajl_val node, container_stats_t * stats) {
+  stats->network_stats = (network_stats_t *) calloc(1, sizeof(network_stats_t));
+  if (stats->network_stats == NULL) {
+    ERROR("docker: get_network_stats: malloc failed!");
+  }
+
+  yajl_val network_node = yajl_tree_get(node, NETWORK_PATH, yajl_t_object);
+  if (!YAJL_IS_OBJECT(network_node)) {
+    ERROR("docker: JSON Error. \n");
+    sfree(stats->network_stats);
+    return;
+  }
+  unsigned int num_interfaces = STATIC_ARRAY_SIZE(network_node->u.object.keys);
+  stats->network_stats->interfaces = (interface_stats_t **)
+      calloc(num_interfaces, sizeof(interface_stats_t *));
+  if (stats->network_stats->interfaces == NULL) {
+    ERROR("docker: get_network_stats: malloc failed!");
+    sfree(stats->network_stats);
+    stats->network_stats = NULL;
+    return;
+  }
+  stats->network_stats->count = num_interfaces;
+  for (int i = 0; i < num_interfaces; i++) {
+    stats->network_stats->interfaces[i] = (interface_stats_t *)
+        calloc(1, sizeof(interface_stats_t));
+    if (stats->network_stats->interfaces[i] == NULL) {
+      ERROR("docker: get_network_stats: malloc failed!");
+      continue;
```

Comment:
What if this for loop executes and none of the calloc()s work. Should we sfree(stats->network_stats) once we get out of the loop?

Also, should stats->network_stats->count be set based on the actual number of interfaces we're able to process successfully?